packages/@glimmer/validator/lib/tracking.ts

```diff
@@ -108,64 +118,122 @@ export function memoizeTracked<T>(
   callback: (...args: unknown[]) => T,
   debuggingContext?: string | false
 ) {
-  let lastValue: T | undefined;
-  let tag: Tag;
-  let snapshot: Revision;
+  let cache = new Cache(callback, debuggingContext);
 
-  let memoized = (...args: unknown[]): T => {
-    if (!tag || !validateTag(tag, snapshot)) {
+  let memoized = (...args: unknown[]) => getValue(cache, ...args);
+
+  (memoized as any)[CACHE_KEY] = cache;
+
+  return memoized;
+}
+
+export function isConstMemo(memoized: Function) {
+  let cache = (memoized as any)[CACHE_KEY] as Cache | undefined;
+
+  return cache === undefined ? false : isConst(cache);
+}
+
+//////////
+
+let DEBUG_CACHE = symbol('DEBUG_CACHE');
+let DEBUG_CACHE_CONTEXT = symbol('DEBUG_CACHE_CONTEXT');
+
+class Cache<T = unknown> {
+  private lastValue: T | undefined;
+  private tag: Tag | undefined;
+  private snapshot: Revision = -1;
+
+  constructor(private fn: (...args: unknown[]) => T, debuggingContext?: string | false) {
+    if (DEBUG) {
+      return ({
+        [DEBUG_CACHE]: this,
+        [DEBUG_CACHE_CONTEXT]: debuggingContext,
+      } as any) as Cache<T>;
+    }
+  }
+
+  static getValue<T>(cache: Cache<T>, ...args: unknown[]): T {
```

Comment:
Why is this a static function? It isn't clear to me (it doesn't use `this`), and it doesn't seem to coincide with the API from the RFC (which is that `getValue` is a stand alone function).

Reply:
`static` allows us to access the `private` fields of an instance, without marking them as public. It's a nifty trick for us to have friend functions without using Symbols.

Reply:
Sure, but:

1. we aren't doing that (no private usage here AFAICT)
2. we only need extra protection in DEBUG builds, where we can already use other techniques (e.g. `WeakMap`/`WeakSet`) to ensure that what we pass back from `createCache` is what we are given here